### PR TITLE
Add Range Interface to Triangle and Quad

### PR DESCRIPTION
**Pull request by fsimonis**
This will change Triangle and Quad so that they fulfil the requirements of the RandomAccessRange model.
The iterators provide access to the vertex coordinates and are all implemented as ReadIterators.

### CODE REVIEW

**Comment by floli** [comment]
Why only for Triangle and Quad, why not for Edge, too?

**Comment by fsimonis** [comment]
The reason is that this PR is the base for the `boost.geomerty` adaptation which models triangles and quads as rings (i.e. random access range of points). In contrast, edges fit the model segment which needs to provide access via a struct specialisation.

So range access to edges is not necessary for the adaptation.
I have to admit that it would be more consistent though.

File: src/mesh/Quad.hpp [hunk 1]
```diff
@@ -160,5 +215,10 @@ inline int Quad::getID() const
   return _id;
 }
 
+BOOST_CONCEPT_ASSERT((boost::RandomAccessIteratorConcept<Quad::iterator>));
```
[the remark this answers is not in the document]
Comment (fsimonis) [reply src/mesh/Quad.hpp]: Lets put them into `vertex.cpp`. This way, the requirement is still visible and the build fails early whilst slowing down only a single TU.
The tests get compiled after the library, so the assert in tests looses its purpose.

**Comment by S0mmerD** [comment]
> Why only for Triangle and Quad, why not for Edge, too?

@floli does have a point here (in my opinion). `*Triangle(...).begin()` is an `Eigen::VectorXd` right now. This is a bit odd, considering that when iterating through a Triangle yields Three vectors (are they edges? are they vertices?). Why not let `begin()`, etc. return a `Vertex`, and let Vertex fulfil the *Point Concept*? Edges are not eligible for the *Polygon Concept*  but they can fulfil the *Linestring Concept*. So even though they are not strictly needed here, why not include them still (I will need them for implementing comparison operator anyway)? So my proposal:

- Vertex: Point Concept
- Edge: Linestring Concept
- Triangle & Quad: Polygon Concept

**Comment by floli** [comment]
I just merged this, while you were commenting. It's fine to continue discussion here, but I think additional commits should go to another pull request.

**Comment by fsimonis** [comment]
@S0mmerD PR #156 contains the reason for that choice. Lets continue the discussion there.

**Pull request merged**